src/core/regimes.rkt

```diff
@@ -173,64 +173,80 @@
-  (define num-candidates (length err-lsts))
-  (define num-points (length (car err-lsts)))
+  (define num-candidates (vector-length err-lsts-vec))
+  (define num-points (length (vector-ref err-lsts-vec 0)))
   (define min-weight num-points)
 
-  (define psums (map (compose partial-sums list->vector) err-lsts))
-  (define can-split? (curry vector-ref (list->vector can-split-lst)))
-
+  ;; Vector version of psums, idk how to inline make-vec-psum yet
+  (define (make-vec-psum lst) 
+    (partial-sums (list->vector lst)))
+  (define vec-psums (vector-map make-vec-psum err-lsts-vec))
+  (define can-split? (curry vector-ref can-split-vec))
   ;; Our intermediary data is a list of cse's,
   ;; where each cse represents the optimal splitindices after however many passes
   ;; if we only consider indices to the left of that cse's index.
   ;; Given one of these lists, this function tries to add another splitindices to each cse.
+  (define (when-cond entry)
+    (can-split? (si-pidx (vector-ref (cse-indices entry)0))))
   (define (add-splitpoint sp-prev)
+    (struct idk (cidx pidx) #:transparent)
     ;; If there's not enough room to add another splitpoint, just pass the sp-prev along.
-    (for/vector #:length num-points ([point-idx (in-naturals)] [point-entry (in-vector sp-prev)])
+    (define result (make-vector (+ num-points)))
+    (for ([point-idx (in-naturals)] [point-entry (in-vector sp-prev)])
       ;; We take the CSE corresponding to the best choice of previous split point.
       ;; The default, not making a new split-point, gets a bonus of min-weight
       (let ([acost (- (cse-cost point-entry) min-weight)] [aest point-entry])
-        (for ([prev-split-idx (in-range 0 point-idx)] [prev-entry (in-vector sp-prev)]
-              #:when (can-split? (si-pidx (car (cse-indices prev-entry)))))
+        (for ([prev-split-idx (in-range 0 point-idx)] 
+              [prev-entry (in-vector sp-prev)]
+              #:when (when-cond prev-entry))
           ;; For each previous split point, we need the best candidate to fill the new regime
           (let ([best #f] [bcost #f])
-            (for ([cidx (in-naturals)] [psum (in-list psums)])
+            (for ([cidx (in-naturals)] [psum (in-vector vec-psums)])
               (let ([cost (- (vector-ref psum point-idx)
                              (vector-ref psum prev-split-idx))])
                 (when (or (not best) (< cost bcost))
                   (set! bcost cost)
                   (set! best cidx))))
             (when (and (< (+ (cse-cost prev-entry) bcost) acost))
               (set! acost (+ (cse-cost prev-entry) bcost))
-              (set! aest (cse acost (cons (si best (+ point-idx 1))
+              (define idk (si best (+ point-idx 1)))
+              (cond ((struct? idk) (set! idk (vector idk))))
+              (set! aest (cse acost (vector-append idk 
                                           (cse-indices prev-entry)))))))
```

Comment:
Ok, look, using `vector-append` here doesn't help, it just forces even more allocation. Instead, _just store the new index_. If you want the full list, you can reconstruct it by following the pointers; we don't need that except at the end.

Reply:
Sorry read this after commenting on the former.

Reply:
I will work on doing this next? 

I figured `vector-append` was expensive but it got the data away from list towards the vector data structure.

Reply:
Ok fine sure

Reply:
As in, separate PR? Ok. Ask for rereview when you want one.